src/fabric/src/fabric_util.erl

```diff
@@ -423,6 +395,103 @@ do_isolate(Fun) ->
             {'$isolerr', Tag, Reason, Stack}
     end.
 
+row_embed_doc(#view_row{} = Row, Doc) ->
+    Row#view_row{doc = Doc};
+row_embed_doc(#{} = Row, Doc) ->
+    Row#{doc => Doc}.
+
+add_if_defined(#{} = Map, Key, Value) ->
+    case Value of
+        undefined -> Map;
+        V -> Map#{Key => V}
+    end.
+
+to_view_row_map(#view_row{key = Key, id = Id, value = Value, doc = Doc, worker = Worker}) ->
+    Row0 = #{},
+    Row1 = add_if_defined(Row0, key, Key),
+    Row2 = add_if_defined(Row1, id, Id),
+    Row3 = add_if_defined(Row2, value, Value),
+    Row4 = add_if_defined(Row3, doc, Doc),
+    add_if_defined(Row4, worker, Worker);
+to_view_row_map(#{} = Row) ->
+    Row.
+
+to_view_row_record(#view_row{} = Row) ->
+    Row;
+to_view_row_record(#{} = Row) ->
+    Id = maps:get(id, Row, undefined),
+    Key = maps:get(key, Row, undefined),
+    Value = maps:get(value, Row, undefined),
+    Doc = maps:get(doc, Row, undefined),
+    Worker = maps:get(worker, Row, undefined),
+    #view_row{id = Id, key = Key, value = Value, doc = Doc, worker = Worker}.
+
+row_get_worker(#view_row{worker = Worker}) ->
+    Worker;
+row_get_worker(#{worker := Worker}) ->
+    Worker;
+row_get_worker(#{}) ->
+    undefined.
+
+row_get_value(#view_row{value = Value}) ->
+    Value;
+row_get_value(#{value := Value}) ->
+    Value;
+row_get_value(#{}) ->
+    undefined.
+
+insert_row(_Dir, Row, []) ->
+    [Row];
+insert_row(fwd, #{id := RowId} = Row, [#{id := HeadId} = Head | Tail] = List) ->
+    case RowId =< HeadId of
+        true -> [Row | List];
+        false -> [Head | insert_row(fwd, Row, Tail)]
+    end;
+insert_row(rev, Row, List) ->
+    lists:reverse(insert_row(fwd, Row, lists:reverse(List))).
+
+merge_row(fwd, #view_row{} = Row, Rows) ->
+    lists:keymerge(#view_row.id, [Row], Rows);
+merge_row(rev, #view_row{} = Row, Rows) ->
+    lists:rkeymerge(#view_row.id, [Row], Rows);
+merge_row(Dir, #{} = Row, Rows) ->
+    insert_row(Dir, Row, Rows).
```

Comment:
Wonder if `lists:merge(Fun, List1, List2)`  would be simpler here? Similar to `how merge_row(Dir, Collation, undefined, Row, Rows0)` works?

In other words since we have to use a custom insert_row call anyway, we could revert to using `lists:merge(fun(RowA, RowB -> case Dir of fwd -> ... ; rev -> ... end end), [Row], Rows)`

Then as long we know how to fetch the Id from the rows they don't even need to of the same type, they could be a mix of maps and records

Reply:
You are probably right.  I was trying to find an alternative solution to `lists:keymerge/2` in the standard Erlang libraries, but I could not come up with anything better.  I noticed that this use case basically becomes an insertion sort and that is what I captured in the PR.  Thanks for the proposal, I like it — I am all for higher-order functions and this sounds like a great chance to utilize them more, even for lists of mixed-format records.